### PR TITLE
[Merged by Bors] - feat (Mathlib/FieldTheory/RatFunc.lean): add five lemmas about coercions between Hahn or Laurent Series and Rational Functions

### PR DESCRIPTION
**Pull request by faenuccio**
Add five lemmas about coercions between Hahn or Laurent Series and Rational Functions

Co-authored-by: María Inés de Frutos Fernández @mariainesdff 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/FieldTheory/RatFunc.lean
Comment (awainverse) [inline Mathlib/FieldTheory/RatFunc.lean:1775]: In two of these three theorems, `single_pow` and `single_zpow`, the left side is an expression `single a b`, and the right side has other functions applied. In `single_inv`, this convention is reversed.
Should there be a convention, possibly even for `simp`, where the expression of the form `single a b` is either consistently the LHS or consistently the RHS?

File: Mathlib/FieldTheory/RatFunc.lean
Comment (awainverse) [inline Mathlib/FieldTheory/RatFunc.lean:184]: Is this comment useful?

**Review by awainverse — COMMENTED** [review]
Looks good, one small convention question.

File: Mathlib/FieldTheory/RatFunc.lean
Comment (riccardobrasca) [inline Mathlib/FieldTheory/RatFunc.lean:822]: Why adding the `by`?

File: Mathlib/FieldTheory/RatFunc.lean
Comment (riccardobrasca) [inline Mathlib/FieldTheory/RatFunc.lean:1710]: Is there a reason `f \neq 0` is not a variable?

File: Mathlib/FieldTheory/RatFunc.lean
Comment (riccardobrasca) [inline Mathlib/FieldTheory/RatFunc.lean:1783]:
```suggestion
theorem single_inv (d : ℤ) {α : F} (hα : α ≠ 0) :
```

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:1710]: Basically that I never remember what is prefered. I have changed it (is it better?)

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:822]: Old tentive thats stayed there, thanks for the catch.

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:184]: No! Good catch, thanks.

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:1775]: Good point. I do not know about other parts of the library, but at least I have aligned them here.

Comment (Vierkantor) [reply Mathlib/FieldTheory/RatFunc.lean:1774]: I agree with the direction that you have aligned them to. Basically, we should see `single` as a homomorphism and try to make its arguments smaller.

File: Mathlib/FieldTheory/RatFunc.lean
Comment (Vierkantor) [inline Mathlib/FieldTheory/RatFunc.lean:817]: Could you define a separate function? That's required for the `@[coe]` annotation which we need for certain tactics. Something like (untested):
```suggestion
@[coe]
def coePolynomial (P : Polynomial K) : RatFunc K := algebraMap _ _ P

instance : Coe (Polynomial K) (RatFunc K) := ⟨coePolynomial⟩
```

File: Mathlib/FieldTheory/RatFunc.lean
Comment (Vierkantor) [inline Mathlib/FieldTheory/RatFunc.lean:1702]: A proof without `erw`/`rfl` so better resistant against library changes:
```suggestion
  simp [RatFunc.coe_def]
```

File: Mathlib/FieldTheory/RatFunc.lean
Comment (Vierkantor) [inline Mathlib/FieldTheory/RatFunc.lean:1797]: All of these are only about `HahnSeries`, right? Could you move them to an appropriate file in `RingTheory/HahnSeries`?

**Review by Vierkantor — COMMENTED** [review]
General remark, not for this PR: should we split this file? It's getting quite long and diverse...

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:817]: Done (and it worked)

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:1774]: Taken the first one.

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:1702]: If you're speaking about the proof of `coe_coe`, the nice `simp [RatFunc.coe_def]` proof does not work for me...

Comment (Vierkantor) [reply Mathlib/FieldTheory/RatFunc.lean:1702]: Hm, it worked for me... Maybe you also need to unfold the new `coePolynomial` definition in `simp`?

Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:1702]: Yep! Good catch.

File: Mathlib/FieldTheory/RatFunc.lean
Comment (faenuccio) [reply Mathlib/FieldTheory/RatFunc.lean:1797]: I would be happy to do so, but I think this is somewhat connected to your comment about splitting the file (which I agree with). My main question comes from the fact that Laurent series are defined as an abbreviation of Hahn Series for the monoid Z and so I tend to consider that everything pertaining to Laurent series actually has to do to Hahn ones. Would it be possible to 
1. Merge this PR as it is (modulo other remarks about proofs/naming/etc)
2. We discuss the rationale beyond splitting this and possibly other files touching Laurent series
3. I open a new PR where I split the file?

**Comment by faenuccio** [comment]
> General remark, not for this PR: should we split this file? It's getting quite long and diverse...

I fully agree and have addressed this point in my final comment about moving some results to the Hahn series file.

File: Mathlib/FieldTheory/RatFunc.lean
Comment (Vierkantor) [reply Mathlib/FieldTheory/RatFunc.lean:1798]: Okay, sounds good to me!

**Review by Vierkantor — APPROVED** [review]
Thanks for your changes, it's looking good to me. I think since you've addressed all the reviewers' remarks you should be ready to merge this, but I'll give them some time before I hit merge for the final complaints.

**Comment by riccardobrasca** [comment]
Splitting the file seems a good idea, thanks!

bors merge

**Pull request closed without merging**
